packages/@react-aria/focus/src/useFocusRing.ts

```diff
@@ -49,7 +56,7 @@ export function useFocusRing(props: FocusRingProps = {}): FocusRingAria {
 
   return {
     isFocused: within ? isFocusWithin : isFocused,
-    isFocusVisible: (within ? isFocusWithin : isFocused) && isFocusVisible,
+    isFocusVisible: (within ? isFocusWithin : isFocused) && isFocusVisibleState,
```

Comment:
I think maybe the problem is that the focus visible state for non-focused components doesn't get updated anymore, which means this boolean returns false. Because the focus visible listener isn't fired when the element receives focus, so the state is stale.

Would it make sense to have a single state rather than two here? And then store the current focus and focus visibility state in a ref instead so it doesn't trigger a re-render and cause performance issues. Something like this?

```js
let state = useRef({
  isFocused: autoFocus,
  isFocusVisible: isFocusVisible()
}).current;
let [isFocusVisible, setFocusVisible] = useState(() => state.isFocused && state.isFocusVisible);

let updateState = () => {
  setFocusVisible(state.isFocused && state.isFocusVisible);
};

useFocusVisibleListener((isFocusVisible) => {
  state.isFocusVisible = isFocusVisible;
  updateState();
});

let onFocusChange = isFocused = {
  state.isFocused = isFocused;
  updateState();
};

let {focusProps} = useFocus({
  isDisabled: within,
  onFocusChange
});

let {focusProps} = useFocusWithin({
  isDisabled: !within,
  onFocusChange
});
```